### PR TITLE
flatten: array of key/value records

### PR DESCRIPTION
**Pull request by mattnibs**
Change flatten to return an array of key/value records where the
key is a string array of the value path and value is the value (any
type).

Part of #3615

### CODE REVIEW

File: runtime/expr/function/flatten.go [hunk 1]
```diff
@@ -11,15 +12,18 @@ import (
 // https://github.com/brimdata/zed/blob/main/docs/language/functions.md#flatten
 type Flatten struct {
 	zcode.Builder
-	keyType zed.Type
-	mapper  *zed.Mapper
-	zctx    *zed.Context
+	keyType    zed.Type
+	mapper     *zed.Mapper
+	entryTypes map[int]zed.Type
```
Comment (nwt) [inline runtime/expr/function/flatten.go:17]: Nit: This obviously works either way, but I think the intention is clearer if it's a map from zed.Type to zed.Type.
```suggestion
	entryTypes map[zed.Type]zed.Type
```

File: runtime/expr/function/flatten.go [hunk 2]
```diff
@@ -40,12 +44,35 @@ func (n *Flatten) Call(ctx zed.Allocator, args []zed.Value) *zed.Value {
...
+			if !ok {
+				var err error
+				typ, err = n.zctx.LookupTypeRecord([]zed.Column{
+					zed.NewColumn("key", n.keyType),
+					zed.NewColumn("value", col.Type),
+				})
+				if err != nil {
+					panic(fmt.Errorf("flatten: error creating record type %w", err))
+				}
+				n.entryTypes[col.Type.ID()] = typ
```
Comment (nwt) [inline runtime/expr/function/flatten.go:70]:
```suggestion
				n.entryTypes[col.Type] = n.zctx.MustLookupTypeRecord([]zed.Column{
					zed.NewColumn("key", n.keyType),
					zed.NewColumn("value", col.Type),
				})
```

File: runtime/expr/function/flatten.go [hunk 3]
```diff
@@ -40,12 +44,35 @@ func (n *Flatten) Call(ctx zed.Allocator, args []zed.Value) *zed.Value {
...
+	return ctx.NewValue(n.zctx.LookupTypeArray(valType), n.Bytes())
+}
+
+func (n *Flatten) collectTypes(cols []zed.Column) []zed.Type {
+	var types []zed.Type
+	for _, col := range cols {
+		if typ := zed.TypeRecordOf(col.Type); typ != nil {
+			for _, typ := range n.collectTypes(typ.Columns) {
+				types = append(types, typ)
+			}
```
Comment (nwt) [inline runtime/expr/function/flatten.go:58]:
```suggestion
			types = append(types, n.collectTypes(typ.Columns)...)
```

**Review by nwt — APPROVED** [review]

File: runtime/expr/function/flatten.go (hunk 1 above)
Comment (mattnibs) [reply runtime/expr/function/flatten.go:17]: Good point.

**Comment by mattnibs** [comment]
I need to update the flatten docs.

File: runtime/expr/function/flatten.go (hunk 2 above)
Comment (mattnibs) [reply runtime/expr/function/flatten.go]: This doesn't set `typ` but I'll fix this.

**Pull request merged**